File: typescript/src/resources/emails.ts
/**
 * Emails resource - Email sending, scheduling, and validation
 */

import type { HttpClient } from '../utils/http';
import type { SendEmailRequest, SendEmailResponse, EmailEvent } from '../types';
import type { RetryConfig } from '../utils/retry';
import { withRetry } from '../utils/retry';

export class EmailsResource {
  constructor(
    private http: HttpClient,
    private retryConfig: RetryConfig
  ) {}

  /**
   * Send a transactional email
   * 
   * @param request - Email send request
   * @returns Send response with message ID and quota info
   * 
   * @example
   * ```typescript
   * const result = await client.emails.send({
   *   from: { email: 'noreply@example.com', name: 'Example' },
   *   to: [{ email: 'user@example.com', name: 'User' }],
   *   content: {
   *     subject: 'Welcome!',
   *     html: '<h1>Welcome to our service!</h1>',
   *     text: 'Welcome to our service!'
   *   }
   * });
   * console.log('Message ID:', result.data.messageId);
   * ```
   */
  public async send(request: SendEmailRequest): Promise<SendEmailResponse> {
    return withRetry(
      () => this.http.post<SendEmailResponse>('/api/v1/emails/send', request),
      this.retryConfig
    );
  }

  /**
   * Send email using a template
   * 
   * @param templateId - Template ID
   * @param to - Recipient(s)
   * @param variables - Template variables
   * @param options - Additional options (from, cc, bcc, themeId)
   * @returns Send response
   * 
   * @example
   * ```typescript
   * const result = await client.emails.sendTemplate(
   *   'tpl_abc123',
   *   [{ email: 'user@example.com' }],
   *   { name: 'John', verificationCode: '123456' },
   *   { from: { email: 'noreply@example.com' } }
   * );
   * ```
   */
  public async sendTemplate(
    templateId: string,
    to: Array<{ email: string; name?: string }>,
    variables: Record<string, string>,
    options: {
      from: { email: string; name?: string };
      cc?: Array<{ email: string; name?: string }>;
      bcc?: Array<{ email: string; name?: string }>;
      replyTo?: { email: string; name?: string };
      themeId?: string;
    }
  ): Promise<SendEmailResponse> {
    const request: SendEmailRequest = {
      from: options.from,
      to,
      cc: options.cc,
      bcc: options.bcc,
      replyTo: options.replyTo,
      content: {
        subject: '', // Will be filled from template
        templateId,
      },
      variables,
      themeId: options.themeId,
    };

    return this.send(request);
  }

  /**
   * Schedule an email for future delivery
   * 
   * @param request - Email send request
   * @param scheduledFor - ISO 8601 timestamp
   * @returns Scheduled email details
   * 
   * @example
   * ```typescript
   * const futureDate = new Date();
   * futureDate.setHours(futureDate.getHours() + 24);
   * 
   * const result = await client.emails.schedule(
   *   {
   *     from: { email: 'noreply@example.com' },
   *     to: [{ email: 'user@example.com' }],
   *     content: { subject: 'Reminder', text: 'Your appointment is tomorrow' }
   *   },
   *   futureDate.toISOString()
   * );
   * ```
   */
  public async schedule(
    request: SendEmailRequest,
    scheduledFor: string
  ): Promise<{ success: true; data: { id: string; scheduledFor: string } }> {
    return withRetry(
      () =>
        this.http.post('/api/v1/emails/schedule', {
          ...request,
          scheduledFor,
        }),
      this.retryConfig
    );
  }

  /**
   * Cancel a scheduled email
   * 
   * @param messageId - Message ID of scheduled email
   * @returns Cancellation confirmation
   */
  public async cancel(messageId: string): Promise<{ success: true }> {
    return withRetry(
      () => this.http.post(`/api/v1/emails/${messageId}/cancel`),
      this.retryConfig
    );
  }

  /**
   * Validate email addresses (syntax and deliverability)
   * 
   * @param emails - Email address(es) to validate
   * @returns Validation results
   * 
   * @example
   * ```typescript
   * const result = await client.emails.validate('user@example.com');
   * console.log('Valid:', result.data.valid);
   * console.log('Deliverable:', result.data.deliverable);
   * ```
   */
  public async validate(
    emails: string | string[]
  ): Promise<{
    success: true;
    data: Array<{
      email: string;
      valid: boolean;
      deliverable: boolean;
      reason?: string;
    }>;
  }> {
    const emailList = Array.isArray(emails) ? emails : [emails];

    if (emailList.length === 1) {
      return withRetry(
        () => this.http.post('/api/v1/emails/validate', { email: emailList[0] }),
        this.retryConfig
      );
    } else {
      return withRetry(
        () => this.http.post('/api/v1/emails/validate/bulk', { emails: emailList }),
        this.retryConfig
      );
    }
  }

  /**
   * Send batch emails (up to 1000 recipients)
   * 
   * @param emails - Array of email send requests
   * @returns Batch send results
   * 
   * @example
   * ```typescript
   * const result = await client.emails.sendBatch([
   *   {
   *     from: { email: 'noreply@example.com' },
   *     to: [{ email: 'user1@example.com' }],
   *     content: { subject: 'Hello 1', text: 'Message 1' }
   *   },
   *   {
   *     from: { email: 'noreply@example.com' },
   *     to: [{ email: 'user2@example.com' }],
   *     content: { subject: 'Hello 2', text: 'Message 2' }
   *   }
   * ]);
   * ```
   */
  public async sendBatch(
    emails: SendEmailRequest[]
  ): Promise<{
    success: true;
    data: { accepted: number; rejected: number; results: Array<{ messageId: string; status: string }> };
  }> {
    if (emails.length === 0) {
      throw new Error('At least one email is required');
    }

    if (emails.length > 1000) {
      throw new Error('Maximum 1000 emails per batch');
    }

    return withRetry(
      () => this.http.post('/api/v1/emails/batch', { emails }),
      this.retryConfig
    );
  }

  /**
   * Get email events for a specific message
   * 
   * @param messageId - Message ID
   * @returns Email events (sent, delivered, opened, clicked, bounced, etc.)
   */
  public async getEvents(messageId: string): Promise<{ success: true; data: EmailEvent[] }> {
    return withRetry(
      () => this.http.get(`/api/v1/emails/${messageId}/events`),
      this.retryConfig
    );
  }
}

File: python/northrelay/resources/emails.py
"""Resource module"""
from __future__ import annotations

"""Emails resource - Email sending, scheduling, and validation"""

from typing import Any, Optional
from datetime import datetime

from northrelay.utils.http import HttpClient
from northrelay.utils.retry import with_retry, RetryConfig
from northrelay.types import (
    EmailAddress,
    SendEmailRequest,
    SendEmailResponse,
)


class EmailsResource:
    """Email sending and management"""

    def __init__(self, http: HttpClient, retry_config: RetryConfig):
        self._http = http
        self._retry_config = retry_config

    async def send(self, request: SendEmailRequest) -> SendEmailResponse:
        """
        Send a transactional email

        Args:
            request: Email send request with from, to, content, etc.

        Returns:
            SendEmailResponse with message_id and quota info

        Raises:
            AuthenticationError: Invalid API key
            ValidationError: Invalid request data
            QuotaExceededError: Email quota exceeded
            RateLimitError: Rate limit exceeded
            ServerError: Server error (5xx)
            NetworkError: Network/connection error

        Example:
            >>> client = NorthRelay(api_key="nr_live_...")
            >>> response = await client.emails.send(
            ...     SendEmailRequest(
            ...         from_={"email": "noreply@example.com", "name": "Example"},
            ...         to=[{"email": "user@example.com"}],
            ...         content={
            ...             "subject": "Welcome!",
            ...             "html": "<h1>Welcome to our service!</h1>",
            ...             "text": "Welcome to our service!",
            ...         },
            ...     )
            ... )
            >>> print(response.message_id)
        """
        async def _send() -> dict[str, Any]:
            # Convert Pydantic model to dict with aliases
            payload = request.model_dump(by_alias=True, exclude_none=True)
            return await self._http.post("/api/v1/emails/send", json=payload)

        response_data = await with_retry(
            _send,
            max_attempts=self._retry_config.max_attempts,
            initial_delay=self._retry_config.initial_delay,
            max_delay=self._retry_config.max_delay,
            exponential_base=self._retry_config.exponential_base,
        )

        return SendEmailResponse(**response_data)

    async def send_template(
        self,
        template_id: str,
        to: list[EmailAddress],
        variables: dict[str, Any],
        *,
        from_: Optional[EmailAddress] = None,
        cc: Optional[list[EmailAddress]] = None,
        bcc: Optional[list[EmailAddress]] = None,
        reply_to: Optional[EmailAddress] = None,
        theme_id: Optional[str] = None,
        tags: Optional[dict[str, str]] = None,
    ) -> SendEmailResponse:
        """
        Send email using a template

        Args:
            template_id: Template ID
            to: List of recipients
            variables: Template variable substitutions
            from_: Sender (optional, uses default if not provided)
            cc: CC recipients (optional)
            bcc: BCC recipients (optional)
            reply_to: Reply-to address (optional)
            theme_id: Brand theme ID (optional)
            tags: Custom tags (optional)

        Returns:
            SendEmailResponse with message_id

        Example:
            >>> await client.emails.send_template(
            ...     template_id="tpl_abc123",
            ...     to=[{"email": "user@example.com", "name": "John"}],
            ...     variables={"name": "John", "verification_code": "123456"},
            ...     from_={"email": "noreply@example.com", "name": "Example"},
            ... )
        """
        request = SendEmailRequest(
            from_=from_ or EmailAddress(email="noreply@example.com"),
            to=to,
            cc=cc,
            bcc=bcc,
            reply_to=reply_to,
            content={"subject": "", "template_id": template_id},  # Subject from template
            variables=variables,
            theme_id=theme_id,
            tags=tags,
        )

        return await self.send(request)

    async def schedule(
        self,
        request: SendEmailRequest,
        scheduled_for: datetime,
    ) -> dict[str, Any]:
        """
        Schedule an email for future delivery

        Args:
            request: Email send request
            scheduled_for: ISO 8601 timestamp for scheduled delivery

        Returns:
            Scheduled email details with schedule_id

        Example:
            >>> from datetime import datetime, timedelta
            >>> future_time = datetime.now() + timedelta(hours=2)
            >>> await client.emails.schedule(request, scheduled_for=future_time)
        """
        request.scheduled_for = scheduled_for

        async def _schedule() -> dict[str, Any]:
            payload = request.model_dump(by_alias=True, exclude_none=True)
            return await self._http.post("/api/v1/emails/schedule", json=payload)

        return await with_retry(
            _schedule,
            max_attempts=self._retry_config.max_attempts,
            initial_delay=self._retry_config.initial_delay,
        )

    async def send_batch(
        self,
        emails: list[SendEmailRequest],
    ) -> dict[str, Any]:
        """
        Send multiple emails in a batch

        Args:
            emails: List of email send requests

        Returns:
            Batch result with batch_id, accepted_count, rejected_count

        Example:
            >>> emails = [
            ...     SendEmailRequest(...),
            ...     SendEmailRequest(...),
            ... ]
            >>> result = await client.emails.send_batch(emails)
            >>> print(f"Sent {result['accepted_count']} of {len(emails)}")
        """
        async def _send_batch() -> dict[str, Any]:
            payload = {
                "emails": [
                    email.model_dump(by_alias=True, exclude_none=True)
                    for email in emails
                ]
            }
            return await self._http.post("/api/v1/emails/batch", json=payload)

        return await with_retry(
            _send_batch,
            max_attempts=self._retry_config.max_attempts,
            initial_delay=self._retry_config.initial_delay,
        )

    async def validate(self, email: str) -> dict[str, Any]:
        """
        Validate an email address

        Args:
            email: Email address to validate

        Returns:
            Validation result with valid, reason, score

        Example:
            >>> result = await client.emails.validate("user@example.com")
            >>> if result["valid"]:
            ...     print("Email is valid")
        """
        async def _validate() -> dict[str, Any]:
            return await self._http.post(
                "/api/v1/emails/validate",
                json={"email": email},
            )

        return await with_retry(_validate)
